### PR TITLE
Issue#14 Front-end

### LINKED ISSUE #14: Handle Array in function arguments

### PR DESCRIPTION
**Pull request by TomBourret**
FIX #14

### CODE REVIEW

File: front-end/src/function/FunctionParam.h [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 class FunctionParam : public VarDeclaration {
 public:
-    FunctionParam(const std::string &name, Type type);
+    FunctionParam(const std::string &name = "param", Type type = INT64_T);
```
Comment (Embraser01) [inline front-end/src/function/FunctionParam.h]: I don't think default values are needed, if necessary, create an empty constructor

Comment (Embraser01) [reply front-end/src/function/FunctionParam.h]: (Accessible only to child classes `protected:`)

File: front-end/src/function/FunctionParamTab.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+#define MAPLECOMPILER_FUNCTIONPARAMTAB_H
+
+#include "FunctionParam.h"
+
+class FunctionParamTab : public FunctionParam {
+
+public:
+    int size;
+
+    explicit FunctionParamTab(const std::string &name = "param", Type type = INT64_T, int size = -1);
```
Comment (Embraser01) [inline front-end/src/function/FunctionParamTab.h]: Same here, no default except for the size

File: front-end/src/visitor/VisitorFunction.cpp [hunk 3]
```diff
@@ -62,27 +63,64 @@ antlrcpp::Any StartVisitor::visitFunctionDefinition(MapleGrammarParser::Function
 antlrcpp::Any StartVisitor::visitTypeList(MapleGrammarParser::TypeListContext *ctx) {
     auto fParams = new vector<FunctionParam *>();
 
-    for (std::size_t i = 0; i != ctx->ID().size(); i++) {
-        const string &name = ctx->ID(i)->getText();
+    for (std::size_t i = 0; i != ctx->argumentTypeVar().size(); i++) {
```
Comment (Embraser01) [inline front-end/src/visitor/VisitorFunction.cpp]: Prefer a more optimized iterator `for (auto&& item: ctx->argumentTypeVar()) {[...]}`

File: front-end/src/maple-parser/MapleGrammar.g4 [hunk 4]
```diff
@@ -86,8 +86,9 @@ functionDefinition : (TYPE | TYPE_VOID) ID '(' (typeList | TYPE_VOID)? ')' block
 returnStatement : 'return' expr SC ;
 blockFunction : '{' declaration* instruction* '}' ;
 argumentList : ((expr ',')* expr)? ;
-typeList : ((TYPE ID ',')* TYPE ID) ;
-
+typeList : ((argumentTypeVar | argumentTypeArray) ',')*  (argumentTypeVar | argumentTypeArray);
+argumentTypeVar : TYPE ID;
```
Comment (Embraser01) [inline front-end/src/maple-parser/MapleGrammar.g4]: Add `argumentType: argumentTypeVar | argumentTypeArray ;`

visitTypeList: `for (item : ctx->argumentType()) list.push(visit(item));`

visitArgumentType: `if (ctx->argumentTypeVar) return (FunctionParam*) visit(ctx->..; return (FunctionParam*) visit(ctx->argumentTypeArray)`

File: front-end/src/visitor/VisitorFunction.cpp [hunk 5]
```diff
@@ -62,27 +63,64 @@ antlrcpp::Any StartVisitor::visitFunctionDefinition(MapleGrammarParser::Function
...
-        currentSymbolTable->insert(name, new Symbol(currentSymbolTable, fParam, true));
         fParams->push_back(fParam);
     }
+
+    for (std::size_t i = 0; i != ctx->argumentTypeArray().size(); i++) {
+
+        auto fParamTab = (FunctionParamTab*) visit(ctx->argumentTypeArray(i));
+
+        fParams->push_back(fParamTab);
+    }
```
Comment (Embraser01) [inline front-end/src/visitor/VisitorFunction.cpp]: remove loop

File: front-end/src/visitor/VisitorFunction.cpp [hunk 6]
```diff
@@ -62,27 +63,64 @@ antlrcpp::Any StartVisitor::visitFunctionDefinition(MapleGrammarParser::Function
...
+
+    auto fParam = new FunctionParamTab(
+            name,
+            getTypeFromString(ctx->TYPE()->getText()),
+            (int) ctx->INTEGER()
+    );
+
+    currentSymbolTable->insert(name, new Symbol(currentSymbolTable, fParam, true));
+
+    return fParam;
```
Comment (Embraser01) [inline front-end/src/visitor/VisitorFunction.cpp:252]: Cast to (FunctionParam*)

**Review by Embraser01 — CHANGES_REQUESTED** [review]

File: front-end/src/maple-parser/MapleGrammar.g4 (hunk 4 above)
Comment (TomBourret) [reply front-end/src/maple-parser/MapleGrammar.g4]: Done

File: front-end/src/visitor/VisitorFunction.cpp (hunk 3 above)
Comment (TomBourret) [reply front-end/src/visitor/VisitorFunction.cpp]: Ok

File: front-end/src/visitor/VisitorFunction.cpp (hunk 5 above)
Comment (TomBourret) [reply front-end/src/visitor/VisitorFunction.cpp]: Done

File: front-end/src/visitor/VisitorFunction.cpp [hunk 7]
```diff
@@ -61,22 +61,22 @@ antlrcpp::Any StartVisitor::visitFunctionDefinition(MapleGrammarParser::Function
...
 
-    for (std::size_t i = 0; i != ctx->argumentTypeArray().size(); i++) {
+    return fParams;
+}
 
-        auto fParamTab = (FunctionParamTab*) visit(ctx->argumentTypeArray(i));
+antlrcpp::Any StartVisitor::visitArgumentType(MapleGrammarParser::ArgumentTypeContext *ctx) {
 
-        fParams->push_back(fParamTab);
+    if (ctx->argumentTypeVar) {
```
Comment (Embraser01) [inline front-end/src/visitor/VisitorFunction.cpp]: Parentethis

**Review by Embraser01 — APPROVED** [review]

**Pull request merged**